src/main/kotlin/org/jitsi/nlj/RtpEncodingDesc.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.nlj
+
+import org.jitsi.nlj.rtp.SsrcAssociationType
+import org.jitsi.nlj.rtp.VideoRtpPacket
+import org.jitsi.nlj.stats.NodeStatsBlock
+
+/**
+ * Keeps track of information specific to an encoded RTP stream.
+ *
+ * @author Jonathan Lennox
+ */
+class RtpEncodingDesc
+@JvmOverloads
+constructor(
+    /**
+     * The primary SSRC for this encoding.
+     */
+    val primarySSRC: Long,
+    /**
+     * The [RtpLayerDesc]s describing the encoding's layers.
+     */
+    initialLayers: Array<RtpLayerDesc> = arrayOf()
+) {
+    /**
+     * The ssrcs associated with this encoding (for example, RTX or FLEXFEC)
+     * Maps ssrc -> type [SsrcAssociationType] (rtx, etc.)
+     */
+    private val secondarySsrcs: MutableMap<Long, SsrcAssociationType> = HashMap()
+
+    fun addSecondarySsrc(ssrc: Long, type: SsrcAssociationType) {
+        secondarySsrcs[ssrc] = type
+    }
+
+    internal var layers = initialLayers
+        set(newLayers) {
+            for (newLayer in newLayers) {
+                /* TODO: indices should be in order, so traverse old and new layers in linear not quadratic time? */
+                for (oldLayer in field) {
+                    if (oldLayer.layerId == newLayer.layerId) {
+                        newLayer.inheritStatistics(oldLayer)
+                        break
+                    }
+                }
+            }
+            field = newLayers
+        }
+
+    /**
+     * @return the "id" of a layer within this source, across all encodings. This is a server-side id and should
+     * not be confused with any encoding id defined in the client (such us the
+     * rid). This server-side id is used in the layer lookup table that is
+     * maintained in [MediaSourceDesc].
+     */
+    fun encodingId(layer: RtpLayerDesc): Long =
+        primarySSRC or (layer.layerId.toLong() shl 32)
```

Comment:
should `RtpLayerDesc` expose a member or function which does the `toLong() shl 32` ?  Or, maybe `RtpEncodingDesc.kt` should define a named extension member on `RtpDescLayer` which does that operation (with a name that describes what it is/how it's different from the raw `layerId` value)